pkg/apis/smith/v1/types.go

```diff
@@ -193,38 +195,48 @@ type Resource struct {
 	// Explicit dependencies.
 	DependsOn []ResourceName `json:"dependsOn,omitempty"`
 
-	Spec runtime.Object `json:"spec"`
+	// TODO nilebox: make "Normal" default type; there are no defaults for Bundle yet?
+	Type ResourceType `json:"type"`
+
+	Spec runtime.Object `json:"spec,omitempty"`
+
+	PluginName string      `json:"pluginName,omitempty"`
+	PluginSpec *PluginSpec `json:"pluginSpec,omitempty"`
 }
 
-// ToUnstructured returns Spec field as an Unstructured object.
-// It makes a copy if it is an Unstructured already.
-func (r *Resource) ToUnstructured() (*unstructured.Unstructured, error) {
-	if u, ok := r.Spec.(*unstructured.Unstructured); ok {
-		return u.DeepCopy(), nil
-	}
-	u, err := unstructured_conversion.DefaultConverter.ToUnstructured(r.Spec)
-	if err != nil {
-		return nil, err
-	}
-	return &unstructured.Unstructured{
-		Object: u,
-	}, nil
+type ResourceType string
+
+const (
+	Normal ResourceType = ""
+	Plugin ResourceType = "plugin"
+)
+
+type PluginSpec struct {
+	ApiVersion string `json:"apiVersion"`
+	Kind       string `json:"kind"`
+	Name       string `json:"name"`
 }
 
 func (r *Resource) UnmarshalJSON(data []byte) error {
-	type resource struct {
-		Name      ResourceName              `json:"name"`
-		DependsOn []ResourceName            `json:"dependsOn"`
-		Spec      unstructured.Unstructured `json:"spec"`
+	var res struct {
```

Comment:
Because `Resource` has a `runtime.Object` field (`Spec`) and there is no way to unmarshal into it. So we use a temporary struct with `Spec unstructured.Unstructured`